src/NetAnalyzers/Core/Microsoft.NetFramework.Analyzers/DoNotUseInsecureDtdProcessingInApiDesign.cs

```diff
@@ -116,395 +102,363 @@ public void AnalyzeSymbol(SymbolAnalysisContext context)
+                OperationBlockStartAnalysisContext context)
             {
-                SyntaxNode node = context.Node;
-                SemanticModel model = context.SemanticModel;
-
-
-                if (SyntaxNodeHelper.GetDeclaredSymbol(node, model) is not IMethodSymbol methodSymbol ||
-                    methodSymbol.MethodKind != MethodKind.Constructor ||
-                    !((!Equals(methodSymbol.ContainingType, _xmlTypes.XmlDocument)) && methodSymbol.ContainingType.DerivesFrom(_xmlTypes.XmlDocument, baseTypesOnly: true)))
+                if (context.OwningSymbol is not IMethodSymbol methodSymbol
+                    || methodSymbol.MethodKind != MethodKind.Constructor
+                    || Equals(methodSymbol.ContainingType, _xmlTypes.XmlDocument)
+                    || !methodSymbol.ContainingType.DerivesFrom(_xmlTypes.XmlDocument, baseTypesOnly: true))
                 {
                     return;
                 }
 
-                bool hasSetSecureXmlResolver = false;
-
-                IEnumerable<SyntaxNode> assignments = _syntaxNodeHelper.GetDescendantAssignmentExpressionNodes(node);
-                foreach (SyntaxNode a in assignments)
+                var hasSetSecureXmlResolver = new ThreadSafeBoolean(false);
+                context.RegisterOperationAction(context =>
                 {
-                    // this is intended to be an assignment, not a bug
-                    if (hasSetSecureXmlResolver = IsAssigningIntendedValueToPropertyDerivedFromType(a,
-                            model,
-                            (s) =>
-                            {
-                                return SecurityDiagnosticHelpers.IsXmlDocumentXmlResolverProperty(s, _xmlTypes);
-                            },
-                            (n) =>
-                            {
-                                return SyntaxNodeHelper.NodeHasConstantValueNull(n, model) ||
-                                    SecurityDiagnosticHelpers.IsXmlSecureResolverType(model.GetTypeInfo(n).Type, _xmlTypes);
-                            },
-                            out bool isTargetProperty))
+                    var assignment = (IAssignmentOperation)context.Operation;
+
+                    if (!hasSetSecureXmlResolver.Value)
                     {
-                        break;
+                        var result = IsAssigningIntendedValueToPropertyDerivedFromType(
+                            assignment,
+                            property =>
+                                SecurityDiagnosticHelpers.IsXmlDocumentXmlResolverProperty(property, _xmlTypes),
+                            operation =>
+                                operation.HasNullConstantValue()
+                                || SecurityDiagnosticHelpers.IsXmlSecureResolverType(operation.Type, _xmlTypes),
+                            out _);
+                        hasSetSecureXmlResolver.Value = result;
```

Comment:
> (Also it's a good question if the desired behavior is to look for at least one secure assignment instead of making sure the last effective assignment is a secure, but addressing that would take a bigger rewrite with dataflow analysis.)

I will let you decide if it's worth opening a ticket for this.